# Lab Assignment Five: Wide and Deep Network Architectures

### Authors
- Juliana Antonio
- Xiaona Hang
- Chuanqi Deng


### 1. Preparation

In [1]:
import pandas as pd
import numpy as np
from copy import deepcopy
from sklearn import metrics as mt
from sklearn.model_selection import train_test_split


df_ad_extra = pd.read_csv('data/Ad_table (extra).csv')
df_ad_extra.info()

df_ad_extra.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 268255 entries, 0 to 268254
Data columns (total 24 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Maker         268255 non-null  object 
 1    Genmodel     268255 non-null  object 
 2    Genmodel_ID  268255 non-null  object 
 3   Adv_ID        268255 non-null  object 
 4   Adv_year      268255 non-null  int64  
 5   Adv_month     268255 non-null  int64  
 6   Color         246380 non-null  object 
 7   Reg_year      268248 non-null  float64
 8   Bodytype      267301 non-null  object 
 9   Runned_Miles  267200 non-null  object 
 10  Engin_size    266191 non-null  object 
 11  Gearbox       268088 non-null  object 
 12  Fuel_type     267846 non-null  object 
 13  Price         267110 non-null  float64
 14  Engine_power  236444 non-null  float64
 15  Annual_Tax    221580 non-null  object 
 16  Wheelbase     240257 non-null  float64
 17  Height        240454 non-null  float64
 18  Widt

Maker  Genmodel  Genmodel_ID   Adv_ID  Adv_year  Adv_month   Color  \
0  Bentley    Arnage         10_1  10_1$$1      2018          4  Silver   
1  Bentley    Arnage         10_1  10_1$$2      2018          6    Grey   
2  Bentley    Arnage         10_1  10_1$$3      2017         11    Blue   
3  Bentley    Arnage         10_1  10_1$$4      2018          4   Green   
4  Bentley    Arnage         10_1  10_1$$5      2017         11    Grey   

   Reg_year Bodytype Runned_Miles  ... Engine_power Annual_Tax Wheelbase  \
0    2000.0   Saloon        60000  ...          NaN        NaN    3116.0   
1    2002.0   Saloon        44000  ...        450.0        315    3116.0   
2    2002.0   Saloon        55000  ...        400.0        315    3116.0   
3    2003.0   Saloon        14000  ...          NaN        NaN    3116.0   
4    2003.0   Saloon        61652  ...          NaN        NaN    3116.0   

   Height   Width  Length  Average_mpg  Top_speed  Seat_num  Door_num  
0  1515.0  2125.0  5390.0          NaN        NaN       5.0       4.0  
1  1515.0  2125.0  5390.0     13.7 mpg    179 mph       5.0       4.0  
2  1515.0  2125.0  5390.0     14.7 mpg    155 mph       5.0       4.0  
3  1515.0  2125.0  5390.0          NaN        NaN       5.0       4.0  
4  1515.0  2125.0  5390.0          NaN        NaN       5.0       4.0  

[5 rows x 24 columns]

In [2]:
data=deepcopy(df_ad_extra)
data.columns = data.columns.str.replace(' ', '')
# Remove columns that are not needed for the model
data.drop(['Genmodel_ID', 'Adv_ID', 'Annual_Tax'], axis=1, inplace=True)

# Convert Columns numeric data
data['Average_mpg'] = data['Average_mpg'].str.extract(r'(\d+\.\d+)').astype(float)
data['Top_speed'] = data['Top_speed'].str.extract(r'(\d+)').astype(float)
data['Engin_size'] = data['Engin_size'].str.extract(r'(\d+\.\d+)').astype(float)
data['Runned_Miles'] = pd.to_numeric(data['Runned_Miles'], errors='coerce')


# Fill categorical columns with the most frequent value (mode)
categorical_cols = data.select_dtypes(include=['object']).columns
data[categorical_cols] = data[categorical_cols].fillna(
    data[categorical_cols].mode().iloc[0])

# Get the unique values in each categorical column
for col in categorical_cols:
    print(
        f'{col} has {len(data[col].unique())} unique values')
data.dropna(inplace=True)
data.reset_index(drop=True, inplace=True)

# Get numerical columns
numerical_cols = data.select_dtypes(include=['float64', 'int64']).columns
data[numerical_cols] = data[numerical_cols].to_numpy().astype(float)

print("=====================================================")
print(f'Numerical columns: {numerical_cols}')
print(f'Categorical columns: {categorical_cols}')

Maker has 88 unique values
Genmodel has 896 unique values
Color has 22 unique values
Bodytype has 18 unique values
Gearbox has 3 unique values
Fuel_type has 13 unique values
Numerical columns: Index(['Adv_year', 'Adv_month', 'Reg_year', 'Runned_Miles', 'Engin_size',
       'Price', 'Engine_power', 'Wheelbase', 'Height', 'Width', 'Length',
       'Average_mpg', 'Top_speed', 'Seat_num', 'Door_num'],
      dtype='object')
Categorical columns: Index(['Maker', 'Genmodel', 'Color', 'Bodytype', 'Gearbox', 'Fuel_type'], dtype='object')


In [3]:
data.info()
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 216069 entries, 0 to 216068
Data columns (total 21 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Maker         216069 non-null  object 
 1   Genmodel      216069 non-null  object 
 2   Adv_year      216069 non-null  float64
 3   Adv_month     216069 non-null  float64
 4   Color         216069 non-null  object 
 5   Reg_year      216069 non-null  float64
 6   Bodytype      216069 non-null  object 
 7   Runned_Miles  216069 non-null  float64
 8   Engin_size    216069 non-null  float64
 9   Gearbox       216069 non-null  object 
 10  Fuel_type     216069 non-null  object 
 11  Price         216069 non-null  float64
 12  Engine_power  216069 non-null  float64
 13  Wheelbase     216069 non-null  float64
 14  Height        216069 non-null  float64
 15  Width         216069 non-null  float64
 16  Length        216069 non-null  float64
 17  Average_mpg   216069 non-null  float64
 18  Top_

Maker Genmodel  Adv_year  Adv_month   Color  Reg_year Bodytype  \
0  Bentley   Arnage    2018.0        6.0    Grey    2002.0   Saloon   
1  Bentley   Arnage    2017.0       11.0    Blue    2002.0   Saloon   
2  Bentley   Arnage    2017.0       12.0    Blue    2002.0   Saloon   
3  Bentley   Arnage    2018.0        8.0   Green    2002.0   Saloon   
4  Bentley   Arnage    2018.0        6.0  Silver    2002.0   Saloon   

   Runned_Miles  Engin_size    Gearbox  ...    Price  Engine_power  Wheelbase  \
0       44000.0         6.8  Automatic  ...  28750.0         450.0     3116.0   
1       55000.0         6.8  Automatic  ...  29999.0         400.0     3116.0   
2       55000.0         6.8  Automatic  ...  24950.0         450.0     3116.0   
3       67000.0         6.8  Automatic  ...  29995.0         450.0     3116.0   
4       52000.0         6.8  Automatic  ...  26990.0         450.0     3116.0   

   Height   Width  Length  Average_mpg  Top_speed  Seat_num  Door_num  
0  1515.0  2125.0  5390.0         13.7      179.0       5.0       4.0  
1  1515.0  2125.0  5390.0         14.7      155.0       5.0       4.0  
2  1515.0  2125.0  5390.0         13.7      179.0       5.0       4.0  
3  1515.0  2125.0  5390.0         13.7      179.0       5.0       4.0  
4  1515.0  2125.0  5390.0         13.7      179.0       5.0       4.0  

[5 rows x 21 columns]

#### <center> Data Descriptions </center>
    
|Feature | Data Type | Description  |
|-------|---------------|----------|
|Maker:| categorical | Automaker name |
|Genmodel:| categorical| Generic Model name|
|Adv_year:| continuous| Advertisement’s creation year|
|Adv_month:| 0-12 months |Advertisement’s creation month|
|color: | categorical| The color of the car|
|Reg_year: | continuous | This car’s first registration/selling year|
|Body_type: | categorical| This car’s body type|
|Runned_miles: | continous | This car’s runned mileage|
|Engin_size: | continous | This car’s engine size|
|Gearbox: | continous | This car’s gearbox |
|Fuel_type: | categorical | This car’s fuel type |
|**Price**: | continous | This car’s selling price| 
|Engine_power: | continous | This car's enging power|
|Wheelbase: | continous | Horizontal distance between the centers of the front and rear wheels|
|Height: | continous | Height of the car|
|Width:  | continous | Width of the car|
|Length: | continous | Length of the car|
|Avg_mpg: | continous | Average miles per gallon |
|Top_speed: | continous | Highest speed car can reach |
|Seat_num: | discrete | This car’s seats number|
|Door_num: | discrete | This car's doors number|



----

In [4]:

import sys

try:
    from sklearn import metrics as mt
    import tensorflow as tf
    from tensorflow import keras

    # Print the versions of TensorFlow, Keras, and Python
    print(f'TensorFlow version: {tf.__version__}')
    # print(f'Keras version: {keras.__version__}')
    print(f'Python version: {sys.version}')

except ImportError as e:
    print(f'Error importing libraries: {e}')
    print('Make sure you have the necessary packages installed (tensorflow, sklearn, etc.)')

except Exception as e:
    print(f'An error occurred: {e}')

TensorFlow version: 2.15.0
Python version: 3.11.5 | packaged by conda-forge | (main, Aug 27 2023, 03:33:12) [Clang 15.0.7 ]


In [5]:
# Create a tensorflow dataset, for ease of use later
df_train, df_test = train_test_split(data, test_size=0.2)
batch_size = 64

def create_dataset_from_dataframe(df_input):
    
    df = df_input.copy()
    labels = df['Price']
    
    df = {key: value.values[:, np.newaxis] for key, value in df_input.items()}
    
    # create a tf.data.Dataset object
    ds = tf.data.Dataset.from_tensor_slices((dict(df), labels))
 
    # enable batching and prefetching
    ds = ds.batch(batch_size)
    ds = ds.prefetch(batch_size)

    return ds

ds_train = create_dataset_from_dataframe(df_train)
ds_test = create_dataset_from_dataframe(df_test)

#### 1.2 Cross-Product

Crossing features such as 'Maker' and 'Bodytype' can uncover relationships between vehicle manufacturers and the body styles of their offerings, providing insights into pricing dynamics influenced by brand reputation and aesthetic preferences. Similarly, crossing 'Gearbox' with 'Fuel_type' can reveal interactions between transmission types and fuel preferences, reflecting performance characteristics and market segmentation. However, caution should be exercised when crossing features with high amount of instances, such as 'Genmodel' with 896 unique instances, as it may introduce computational complexity and increase the risk of overfitting. Therefore, prioritizing cross-product combinations based on their relevance to the domain and balancing computational constraints is essential for effectively capturing interactions while mitigating the potential drawbacks of model complexity.

In [6]:

# FeatureSpace to make data discrete and normalization

# Numerical columns: Index(['Adv_year', 'Adv_month', 'Reg_year', 'Runned_Miles', 'Engin_size',
#        'Price', 'Engine_power', 'Wheelbase', 'Height', 'Width', 'Length',
#        'Average_mpg', 'Top_speed', 'Seat_num', 'Door_num'],
#       dtype='object')
# Categorical columns: Index(['Maker', 'Genmodel', 'Color', 'Bodytype', 'Gearbox', 'Fuel_type'], dtype='object')


from tensorflow.keras.utils import FeatureSpace

# lump everything together, and concatenate
feature_space = FeatureSpace(
    features={
        # Categorical feature encoded as string
        "Maker": FeatureSpace.string_categorical(num_oov_indices=0),
        "Genmodel": FeatureSpace.string_categorical(num_oov_indices=0),
        "Color": FeatureSpace.string_categorical(num_oov_indices=0),
        "Bodytype": FeatureSpace.string_categorical(num_oov_indices=0),
        "Gearbox": FeatureSpace.string_categorical(num_oov_indices=0),
        "Fuel_type": FeatureSpace.string_categorical(num_oov_indices=0),
        
        # Numerical features to make discrete, make into integer with max val
        "Adv_year": FeatureSpace.float_discretized(num_bins=10),
        "Adv_month": FeatureSpace.float_discretized(num_bins=12),
        "Reg_year": FeatureSpace.float_discretized(num_bins=10),
        
        # Numerical features to normalize (normalization will be learned)
        # learns the mean, variance, and if to invert (3 parameters)
        "Runned_Miles": FeatureSpace.float_normalized(),
        "Engin_size": FeatureSpace.float_normalized(),
        "Price": FeatureSpace.float_normalized(),
        "Engine_power": FeatureSpace.float_normalized(),
        "Wheelbase": FeatureSpace.float_normalized(),
        "Height": FeatureSpace.float_normalized(),
        "Width": FeatureSpace.float_normalized(),
        "Length": FeatureSpace.float_normalized(),
        "Average_mpg": FeatureSpace.float_normalized(),
        "Top_speed": FeatureSpace.float_normalized(),
        "Seat_num": FeatureSpace.float_normalized(),
        "Door_num": FeatureSpace.float_normalized(),
    },
# Specify feature cross with a custom crossing dim
    crosses=[
        FeatureSpace.cross(feature_names=('Maker', 'Bodytype'), crossing_dim=88 * 18),
        FeatureSpace.cross(feature_names=('Fuel_type', 'Gearbox'), crossing_dim=13 * 3),
    ],
    output_mode="concat", # can also be a dict, processed internally
)


# Run on the data
# create a version of the dataset that can be iterated without labels
train_ds_with_no_labels = ds_train.map(lambda x, _: x)  
# the adapt function allows the model to learn one-hot encoding sizes
feature_space.adapt(train_ds_with_no_labels) # inititalize the feature map to this data


# now define a preprocessing operation that returns the processed features
preprocessed_ds_train = ds_train.map(lambda x, y: (feature_space(x), y), 
                                     num_parallel_calls=tf.data.AUTOTUNE)
# run it so that we can use the pre-processed data
preprocessed_ds_train = preprocessed_ds_train.prefetch(tf.data.AUTOTUNE)

# Run on the test set
preprocessed_ds_test = ds_test.map(lambda x, y: (feature_space(x), y), num_parallel_calls=tf.data.AUTOTUNE)
preprocessed_ds_test = preprocessed_ds_test.prefetch(tf.data.AUTOTUNE)


#### 1.3 Metric(s) to use to evaluate algorithm’s performance

For evaluating the algorithm's performance in predicting vehicle prices, we will choose Mean Absolute Error (MAE) and Root Mean Squared Error (RMSE) as evaluation metrics. MAE provides a straightforward measure of the average absolute deviation between predicted and actual prices, which is valuable for assessing the overall accuracy of price predictions in terms of magnitude. RMSE complements MAE by penalizing larger errors more heavily, making it sensitive to outliers and providing a more comprehensive assessment of prediction accuracy. Since the primary objective is to minimize prediction errors and ensure accurate pricing, both MAE and RMSE are appropriate metrics that align well with the business case for the task of vehicle price prediction.

In [7]:
# One-hot encoding for the categorical columns
from tensorflow.keras.layers import Embedding, Flatten, Concatenate

def setup_embedding_from_categorical(feature_space, col_name):

    # which is the same as the number of categories
    N = len(feature_space.preprocessors[col_name].get_vocabulary())
    
    # get the output from the feature space, which is input to embedding
    x = feature_space.preprocessors[col_name].output
    
    # use an embedding to deal with integers from feature space
    x = Embedding(input_dim=N, 
                  output_dim=int(np.sqrt(N)), 
                  input_length=1, name=col_name+'_embed')(x)
    
    x = Flatten()(x) # get rid of that extra dimension (for time of embedding)
    
    return x # return the tensor the embedding layer for use in the model

In [8]:
# need to use unprocessed features here, to gain access to each output
dict_inputs = feature_space.get_inputs()

all_branch_outputs = []

# for each feature, setup the embedding
for col in categorical_cols:
    
    # get the output rensor from embedding layer
    x = setup_embedding_from_categorical(feature_space, col)
    all_branch_outputs.append(x)

# for each numerical feature, add it in after embedding
for col in numerical_cols:
    x = feature_space.preprocessors[col].output
    x = tf.cast(x, float) # make sure it is a float
    all_branch_outputs.append(x)

# now concatenate the outputs and add a fully connected layer
final_branch = Concatenate(name='concat_1')(all_branch_outputs)
# final_branch = Dense(units=1,
#                      activation='sigmoid', 
#                      name='combined')(final_branch)



#### 1.4 Dividing Training and Testing Set

We will use shuffle splits to divide the data into training and testing sets for regression, as it ensures random sampling without requiring stratification. This method is appropriate for regression tasks because it provides unbiased estimates of model performance by repeatedly shuffling and splitting the data. It mirrors real-world scenarios where models need to generalize well to unseen data, making it a practical choice for evaluating model performance.

In [9]:
from sklearn.model_selection import ShuffleSplit
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression

# Initialize shuffle splits
shuffle_split = ShuffleSplit(n_splits=10, test_size=0.2, random_state=42)

# Initialize a model (e.g., Linear Regression)
clf = LinearRegression()

# Perform cross-validation
for fold, (train_idx, test_idx) in enumerate(shuffle_split.split(data)):
    # Split data into training and testing sets
    train_data = data.iloc[train_idx]
    test_data = data.iloc[test_idx]
    
    # Separate features and target variable
    X_train, y_train = train_data.drop('Price', axis=1), train_data['Price']
    X_test, y_test = test_data.drop('Price', axis=1), test_data['Price']
    
    # Train the model
    clf.fit(X_train, y_train)
    
    # Make predictions on the test set
    predictions = clf.predict(X_test)
    
    # Evaluate model performance using mean squared error
    mse = mean_squared_error(y_test, predictions)
    print(f"Fold {fold+1} - Mean Squared Error: {mse}")

ValueError: could not convert string to float: 'Volvo'